src/Interactive/Host/Interactive/Core/InteractiveHost.cs

```diff
@@ -413,25 +369,38 @@ private async Task<InitializedRemoteService> TryGetOrCreateRemoteServiceAsync()
             return default;
         }
 
-        private async Task<TResult> Async<TResult>(Action<Service, RemoteAsyncOperation<TResult>> action)
+        private async Task<TResult> Async<TResult>(string targetName, params object[] arguments)
         {
+            var initializedService = await TryGetOrCreateRemoteServiceAsync().ConfigureAwait(false);
+            TaskCompletionSource<TResult> _completion = new TaskCompletionSource<TResult>();
+            EventHandler _processExitedHandler = new EventHandler((_, __) =>
+            {
+                initializedService.ServiceOpt.Process.Exited -= _processExitedHandler;
+                _completion.TrySetResult(default(TResult));
+            });
+
             try
             {
-                var initializedService = await TryGetOrCreateRemoteServiceAsync().ConfigureAwait(false);
-                if (initializedService.ServiceOpt == null)
+                initializedService.ServiceOpt.Process.Exited += _processExitedHandler;
+                if (!initializedService.ServiceOpt.Process.IsAlive())
                 {
-                    return default;
+                    initializedService.ServiceOpt.Process.Exited -= _processExitedHandler;
+                    _completion.TrySetResult(default(TResult));
                 }
 
-                return await new RemoteAsyncOperation<TResult>(initializedService.ServiceOpt).ExecuteAsync(action).ConfigureAwait(false);
+                return await initializedService.ServiceOpt._jsonRpc.InvokeAsync<TResult>(targetName, arguments).ConfigureAwait(false);
             }
-            catch (Exception e) when (FatalError.Report(e))
+            catch (RemotingException) when (!initializedService.ServiceOpt.Process.IsAlive())
             {
-                throw ExceptionUtilities.Unreachable;
+                // the operation might have terminated the process:
+                initializedService.ServiceOpt.Process.Exited -= _processExitedHandler;
+                _completion.TrySetResult(default(TResult));
+                return await _completion.Task.ConfigureAwait(false);
             }
         }
 
-        private static async Task<TResult> Async<TResult>(RemoteService remoteService, Action<Service, RemoteAsyncOperation<TResult>> action)
+        // (miziga) unused method - delete?
+        /*private static async Task<TResult> Async<TResult>(RemoteService remoteService, Action<Service, RemoteAsyncOperation<TResult>> action)
```

Comment:
>private [](start = 10, length = 7)

We want to keep this method. It should do the same as the other Async, except not initialize the service. You can actually move most of the code from the above method to this Async method to share it.